cortex-ar/src/critical_section.rs

```diff
@@ -1,30 +1,147 @@
-        // the i bit means "masked"
-        let was_active = !crate::register::Cpsr::read().i();
-        crate::interrupt::disable();
-        atomic::compiler_fence(atomic::Ordering::SeqCst);
-        was_active
+    /// Indicates the critical section was entered with interrupts on
+    pub const INT_ON: u8 = 0;
+
+    /// Indicates the critical section was entered with interrupts off
+    pub const INT_OFF: u8 = 1;
+
+    #[cfg(feature = "critical-section-single-core")]
+    unsafe impl critical_section::Impl for SingleCoreCriticalSection {
+        unsafe fn acquire() -> critical_section::RawRestoreState {
+            use core::sync::atomic;
+            // the i bit means "masked"
+            let was_active = !crate::register::Cpsr::read().i();
+            crate::interrupt::disable();
+            atomic::compiler_fence(atomic::Ordering::SeqCst);
+            if was_active {
+                INT_ON
+            } else {
+                INT_OFF
+            }
+        }
+
+        unsafe fn release(was_active: critical_section::RawRestoreState) {
+            use core::sync::atomic;
+            // Only re-enable interrupts if they were enabled before the critical section.
+            if was_active == INT_ON {
+                atomic::compiler_fence(atomic::Ordering::SeqCst);
+                // Safety: This is OK because we're releasing a lock that was
+                // entered with interrupts enabled
+                unsafe {
+                    crate::interrupt::enable();
+                }
+            }
+        }
     }
+}
+
+#[cfg(feature = "critical-section-multi-core")]
+mod multi_core {
+    struct MultiCoreCriticalSection;
+
+    critical_section::set_impl!(MultiCoreCriticalSection);
+
+    /// The default value for our spin-lock
+    pub const UNLOCKED: u32 = 0xFFFF_FFFF;
+
+    /// Indicates the critical section was entered with interrupts on, and the spin-lock unlocked
+    pub const INT_ON_UNLOCKED: u8 = 0;
+
+    /// Indicates the critical section was entered with interrupts off, and the spin-lock locked (by us)
+    pub const INT_OFF_LOCKED: u8 = 1;
+
+    /// Indicates the critical section was entered with interrupts off, and the spin-lock unlocked
+    pub const INT_OFF_UNLOCKED: u8 = 2;
+
+    pub static CORE_SPIN_LOCK: core::sync::atomic::AtomicU32 =
+        core::sync::atomic::AtomicU32::new(UNLOCKED);
+    unsafe impl critical_section::Impl for MultiCoreCriticalSection {
+        unsafe fn acquire() -> critical_section::RawRestoreState {
+            use core::sync::atomic;
+
+            // the i bit means "masked"
+            let was_active = !crate::register::Cpsr::read().i();
+            crate::interrupt::disable();
+
+            let core_id = crate::asm::core_id();
+
+            let locked_already = loop {
+                match CORE_SPIN_LOCK.compare_exchange(
+                    UNLOCKED,
+                    core_id,
+                    atomic::Ordering::Acquire,
```

Comment:
I am trying to understand why this does not have to be `AcqRel` .. is it because of the compiler fence after the loop?

Reply:
Oh, I'll definitely read that. I read Jons Rust for Rustacean book about that, but its a very tough subject..